### ISSUE TITLE
Cannot set permission nodes for sub-commands.

### ISSUE THREAD

**Issue by zedwick** [JorelAli/1.13-Command-API#26]
The way we register sub-commands allows for a PermissionNode to be specified, which I intuit to mean that we can set PermissionNodes per executable argument chain, however that doesn't seem to be the case as the specified PermissionNode never yields a permission node in the tree, unless set on the root command. Ideally this would allow for sub-nodes to not tab-complete without set permissions, something we cannot do by checking for the node in our executed code.

Can you confirm if this is as intended or a bug?

V 1.6 with Paper 1.13.2 build 444.

**Comment by JorelAli** [comment]
I don't quite get the question you're asking - could you give me an example of a subcommand and permission registration?

Despite this, I am very certain that I did not intend this.

**Comment by zedwick** [comment]
`LinkedHashMap<String, Argument> library_arguments = new LinkedHashMap<>();
        
        CommandAPI.getInstance().register("library", new CommandPermission("booklibrary.command.library"), library_arguments, (sender, args) -> {
            // /library
        });
        
        library_arguments.put("get", new LiteralArgument("get"));
        library_arguments.put("book_id", new IntegerArgument());
        
        CommandAPI.getInstance().register("library", new CommandPermission("booklibrary.command.get"), library_arguments, (sender, args) -> {
            // /library get <int>
        });
    }`

According to luckperms' `/lp tree` command, there is no `booklibrary.command.get` permission node, and `/library get` is available for tab-completion when the player does not have that permission node.

**Comment by JorelAli** [comment]
So, this issue is caused by a fundamental flaw in the implementation of the CommandAPI. Despite the fact that this registration "works" (i.e. it registers with no issues and runs normally - even the output command dispatcher graph is correct), the one flaw arises due to the fact you "cannot register a command twice".

This is entirely down to a mistake in the implementation of the CommandAPI (or rather, the fact I didn't notice this earlier), and such a fix would require a very in-depth re-writing of the implementation of sub-commands.

For such a "fix" to be found, **information on whether a command is going to have a sub-command needs to be known BEFORE you begin registering the first command.** For example, I need to know that `/cmd blah` is going to be registered after `/cmd`, except currently there is no method of knowing such a thing.

---
Alternatively, the command registration could be deferred until the next command is registered OR the server has finished loading. The downsides with this is that it conflicts with function commands (Commands which use Minecraft's Functions) which must be registered earlier in the server's startup ([as stated here](https://github.com/JorelAli/1.13-Command-API/blob/master/v1.6%20Documentation.md#function-command-registration))

---
ALSO: (And probably the most important thing which I should have stated earlier):
### **It is impossible to register a subcommand with different permissions.**
Based on the server's implementation of how commands and permissions work, this is simply not possible. This is because _each command must have exactly one permission._ The auto-complete, however, doesn't depend on permissions and has its own "setup" which depends on some predicate which is determined when the player's permissions are updated (i.e. if you op a player, or give them a new permission*)

*This isn't very well supported by Bukkit and relogs are required to update these predicates.

---
In short, **I don't believe there's a viable method of actually achieving this**. You either have the main command and all sub-commands have the same permission, or you handle it manually in the command executor's implementation. For example:

```java
LinkedHashMap<String, Argument> arguments = new LinkedHashMap<>();
CommandAPI.getInstance().register("mycommand", arguments, (sender, args) -> {
    if(sender.hasPermission("mypermission")) {
        //code
    }
});
```

(Essentially this is a "won't fix" for me)

**Comment by Minenash** [comment]
This is highly unfortunate, as my server's current command manager allows for permissions on the sub-command level. I was going to attempt to integrate it with this, however not having sub-command permissions is just not possible for our server. I could easily make a command executor wrapper that handles permissions upon execute, but the biggest thing is we don't want sub commands popping up if a player doesn't have access to it :/

> The auto-complete, however, doesn't depend on permissions and has its own "setup"

Would this mean it'd be possible to restrict which sub-commands could be auto completed based on the permission?

**Comment by JorelAli** [comment]
@Minenash 
> Would this mean it'd be possible to restrict which sub-commands could be auto completed based on the permission?
In theory, yes. In practice, it's going to be a lot less elegant of a solution than you'd expect. I've reopened the issue - I've theorized a potential way to fix this, but it's going to require a lot of time to implement (if successful that is).

**Comment by JorelAli** [comment]
As of [this commit](https://github.com/JorelAli/1.13-Command-API/commit/00b9ef3d8afe84c554a6fe06bd548bb37db1c4b8) I've managed to get something to work!

Given the following example:

```
/permissiontest
/permissiontest <hello/world>
```

```java
LinkedHashMap<String, Argument> arguments = new LinkedHashMap<>();

CommandAPI.getInstance().register("permissiontest", CommandPermission.fromString("custompermission"), arguments, (sender, args) -> {
	sender.sendMessage("test1 success!");
});

arguments.put("values", new SuggestedStringArgument("hello", "world"));
CommandAPI.getInstance().register("permissiontest", CommandPermission.fromString("custompermission2"), arguments, (sender, args) -> {
	sender.sendMessage("test2 success!");
});
```

In order to run the `/permissiontest` command, the user MUST have the `custompermission` permission. If they only have `custompermission2`, they will NOT be able to run `/permissiontest`.

HOWEVER, if the user ONLY has `custompermission`, they will NOT be able to _see_ the suggested values "hello" and "world", unless they have the `custompermission2` permission.

-----

However, I don't believe I'll be able to extend this functionality to `LiteralArguments`, as they are "part of the command", whereas other arguments are "suggested". **If you happen to know of a vanilla Minecraft command that has achieved this, do let me know!**

**Comment by Minenash** [comment]
Say we have the commands `/test`, `/test <ex>`, and `/test <ex> <player>` Let's say the first one has permission `A`, the second has permission `B`, and the third has permission `C`. The first only need `A`, and the second needs both `A` and `B`. Does the third one need `A`, `B`, and `C` or only `A` and `C`

**Comment by JorelAli** [comment]
> Doesn't `/execute` do that? If I remember correctly, you must use one of the key terms (`if`, `until`, `store`, ect); if you put anything else (where a key term should be), it flags it as wrong. Wouldn't this be literal behavior instead of suggested?

The `/execute` command, although it does have literals, only has one sole permission (OP level `2` to be precise (see [op-permission-level](https://minecraft.gamepedia.com/Server.properties) for more info)). In short, you're either op and can run the `/execute` command, or you're not op and cannot run the `/execute` command. Regardless, there's no such system (existingly built into vanilla) such that a command has its subcommands (literals/arguments) limited based on your permissions.

-----

> Quick question, say we have the commands `/test`, `/test <ex>`, and `/test <ex> <player>` Let's say the first one has permission `A`, the second has permission `B`, and the third has permission `C`. The first only need `A`, and the second needs both `A` and `B`. Does the third one need `A`, `B`, and `C` or only `A` and `C`

So maybe I was a little vague about permissions. Permit me to go a little bit offtopic for a minute. Firstly, the statement I mentioned earlier in the thread still holds:
### It is impossible to register a subcommand with different permissions.
In fact, this has never existed as an option before - it's always been implemented by you, the developer.

**Bukkit never had subcommands as a feature**
In the [`plugin.yml` documentation](https://bukkit.gamepedia.com/Plugin_YAML), when you register commands, you don't register the subcommands - only the top level command (i.e. the string that is next to the / when executing a command). Also, you can register a permission to said command, but _only one_ permission.

**Vanilla Minecraft does not support subcommands as features**
The CommandAPI is closer to Vanilla Minecraft than Bukkit/Spigot in its implementation, but tries to work seamlessly to hide the complications that the Vanilla Minecraft system represents. In Vanilla Minecraft, commands are registered under a command map, which links to exactly one permission node. Hence, subcommands don't have alternate permissions compared to the parent command.

**We've been circumnavigating the situation all along**
So, given that the two statements above hold, how have we been able to use subcommands with various permissions all this time? Because you, the developer, implement them.

```java
public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
    if(args.length == 0) {
        if(sender.hasPermission(MYPERM) {
            //display help or something
        } else {
            //say you don't have permission to run this command
        }
    } else if(args.length == 1) {
        if(sender.hasPermission(MYPERM.Subcommand) {
            //code
        } else {
            //say you don't have permission to run this command
        }
    } //more stuff with more arg lengths
}
```

**What the CommandAPI can do**
The CommandAPI is not able to implement the subcommand permissions that you, the developer, would normally implement. HOWEVER, it _IS_ possible (as shown a day ago) that there exists a method to hide the suggestions of various arguments to the user.

```
/mycommand
/mycommand <hello/world>
```

Say `/mycommand` requires the permission "perm" to run the command. This is registered via the CommandAPI with no problems. Say, given this new breakthrough, you register the second command which is a `SuggestedStringArgument` containing the words "hello" and "world". You've registered it so it requires the permission "otherperm".

| Only `perm`                                                 | Only `otherperm`          | Both `perm` and `otherperm`                          |
| ----------------------------------------------------------- | ------------------------- | ---------------------------------------------------- |
| Can run both commands, but cannot see the suggested list | Cannot run either command | Can run both commands and can see the suggested list |

**So, to answer your question**
If you had permission `A`, you'd be able to run `/test`, `/test <ex>`, and `/test <ex> <player>`
You'd be able to run the third one with `A`, `B`, and `C` with no problems. If the argument supplies a suggested list, you'd still be able to run the command with only `A` and `C`, but you wouldn't see the suggested list from `<ex>`

**Suggested Arguments**

Only perm | Only otherperm | Both perm and otherperm
-- | -- | --
Can run both commands, but cannot see the suggested list | Cannot run either command | Can run both commands and can see the suggested list

**Comment by Minenash** [comment]
Thanks.
For the record, the question I asked wasn't about running it, but about the suggestions.
So this line was the answer I was looking for `If the argument supplies a suggested list, you'd still be able to run the command with only A and C, but you wouldn't see the suggested list from <ex>` I already got that running it only needs the first perm.

**Comment by Minenash** [comment]
Sounds good 👍

**Comment by JorelAli** [comment]
This has been implemented in [this commit](https://github.com/JorelAli/1.13-Command-API/commit/8595742eca6ffc5dc89a3fa6dab77f965fccbac0) and will be available in the upcoming 1.8 release.

**Comment by JorelAli** [comment]
@Minenash This actually causes an issue, outlined in #35:
Simply put, **if you suggest something, you cannot show the hover text**.

This is a major issue with the CommandAPI as the 1.13 command interface with the hover text plays a key role in the improvement of commands. It allows you to run commands without knowing the entire syntax of the command - you just get reminded of it whilst you type it. In short, I definitely want to keep the hover text feature present.

The only workaround that I know that can resolve this is to limit suggestions if you don't have a permission solely to SuggestedStrings and DynamicSuggestedStrings. If you apply a permission check for other argument types (which do not use suggestions, such as regular Strings, Numbers etc.), it overrides its suggestions (because that's how permission checks are currently implemented), regardless of whether it has suggestions or not and thus, removes the hover text.

**Comment by JorelAli** [comment]
@zedwick @Minenash 
As of [this commit](https://github.com/JorelAli/1.13-Command-API/commit/e2d7160dc1689d110a3165b1e3caf1ebf9090a8d), I have confirmed the possibility of adding permission nodes to subcommands. This feature will be present in the release of v1.8.

**Comment by JorelAli** [comment]
![gif](https://user-images.githubusercontent.com/4613171/50519032-d1afec00-0ab0-11e9-9f94-38be86300e32.gif)

Subcommand permissions have been implemented in 1.8

**Comment by Minenash** [comment]
Thanks ! :D
